java-client/src/main/java/com/cumulocity/sdk/client/notification/SubscriberImpl.java

```diff
@@ -573,13 +588,64 @@ public boolean equals(Object o) {
 
             SubscriptionRecord that = (SubscriptionRecord) o;
 
-            if (id != null ? !id.equals(that.id) : that.id != null) return false;
-            if (listener != null ? !listener.equals(that.listener) : that.listener != null) return false;
-            return subscribeOperationListener != null ? subscribeOperationListener
-                    .equals(that.subscribeOperationListener) : that.subscribeOperationListener == null;
+            if (!Objects.equals(id, that.id)) return false;
+            if (!Objects.equals(listener, that.listener)) return false;
+            return Objects.equals(subscribeOperationListener, that.subscribeOperationListener);
         }
+    }
 
+    private final class SubscriptionWatcher {
+        private final ScheduledExecutorService executorService;
 
-    }
+        public SubscriptionWatcher() {
+            ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                    .setNameFormat("CumulocitySubscriptionWatcher-scheduler-%d")
+                    .build();
+            executorService = Executors.newSingleThreadScheduledExecutor(threadFactory);
+        }
 
+        public void start() {
+            executorService.scheduleWithFixedDelay(
+                    checkSubscriptionsRunner(),
+                    SUBSCRIPTION_WATCHER_DELAY,
+                    SUBSCRIPTION_WATCHER_DELAY,
+                    TimeUnit.SECONDS);
+        }
+
+        public void stop() {
+            executorService.shutdown();
+        }
+
+        private Runnable checkSubscriptionsRunner() {
+            return () -> {
+                log.debug("Running watcher to check subscriptions");
+                Set<SubscriptionRecord> allSubscriptions = new HashSet<>(subscriptions);
+                allSubscriptions.addAll(pendingSubscriptions);
+                allSubscriptions.forEach(this::checkSubscription);
+            };
+        }
+
+        private void checkSubscription(SubscriptionRecord subRec) {
+            ClientSessionChannel channel = getChannel(subRec.getId());
+            if (channel.getSubscribers().isEmpty()) {
+                log.warn("{} bayeux channel {} has no client subscribers",
+                        pendingSubscriptions.contains(subRec) ? "Pending" : "Subscribed", subRec.getId());
+                reSubscribe(subRec);
+            } else {
+                log.debug("Bayeux channel {} has {} client subscriptions (OK)",
+                        subRec.getId(), channel.getSubscribers().size());
+            }
+        }
+
+        private void reSubscribe(SubscriptionRecord subRec) {
+            if (session.isConnected()) {
+                log.info("Trying to subscribe channel {}", subRec.getId());
+                subscribe(subRec.getId(), subRec.subscribeOperationListener, subRec.listener, true);
```

Comment:
can we hardcode `autoRetry=true`?
What about moving this flag to the `SubscriptionRecord` and handle it also here?